src/nxt_otel.c

```diff
@@ -0,0 +1,356 @@
+    nxt_otel_state_transition(r->otel, NXT_OTEL_UNINIT_STATE);
+}
+
+
+static void
+nxt_otel_error(nxt_task_t *t, nxt_http_request_t *r)
+{
+    // purposefully not using state transition helper
+    r->otel->status = NXT_OTEL_UNINIT_STATE;
+    nxt_log(t, NXT_LOG_ERR, "otel error condition");
+    // if r->otel->trace it WILL leak here.
+    // TODO Phase 2: drop trace without sending it somehow?
+}
+
+
+inline void
+nxt_otel_test_and_call_state(nxt_task_t *t, nxt_http_request_t *r)
+{
+    if (r->otel == NULL) {
+        return;
+    }
+
+    switch (r->otel->status) {
+    case NXT_OTEL_UNINIT_STATE:
+      return;
+    case NXT_OTEL_INIT_STATE:
+        nxt_otel_trace_and_span_init(t, r);
+        break;
+    case NXT_OTEL_HEADER_STATE:
+        nxt_otel_span_add_headers(t, r);
+        break;
+    case NXT_OTEL_BODY_STATE:
+        nxt_otel_span_add_body(r);
+        break;
+    case NXT_OTEL_COLLECT_STATE:
+        nxt_otel_span_collect(t, r);
+        break;
+    case NXT_OTEL_ERROR_STATE:
+        nxt_otel_error(t, r);
+        break;
+    }
+}
+
+
+nxt_int_t
+nxt_otel_parse_traceparent(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    nxt_http_request_t *r;
+    char               *copy;
+
+    /* For information on parsing the traceparent header:
+     * https://www.w3.org/TR/trace-context/#traceparent-header
+     * A summary of the traceparent header value format follows:
+     * Traceparent: "$a-$b-$c-$d"
+     *   a. version (2 hex digits) (ff is forbidden)
+     *   b. trace_id (32 hex digits) (all zeroes forbidden)
+     *   c. parent_id (16 hex digits) (all zeroes forbidden)
+     *   d. flags (2 hex digits)
+     */
+
+    r = ctx;
+    if (field->value_length != NXT_OTEL_TRACEPARENT_LEN) {
+        goto error_state;
+    }
+
+    /* strsep is destructive so we make a copy of the field
+     */
+    copy = nxt_mp_zalloc(r->mem_pool, field->value_length + 1);
+    if (copy == NULL) {
+        goto error_state;
+    }
+    memcpy(copy, field->value, field->value_length);
+
+    r->otel->version = (u_char *) strsep(&copy, "-");
+    r->otel->trace_id = (u_char *) strsep(&copy, "-");
+    r->otel->parent_id = (u_char *) strsep(&copy, "-");
+    r->otel->trace_flags = (u_char *) strsep(&copy, "-");
+
+    if (r->otel->version     == NULL ||
+        r->otel->trace_id    == NULL ||
+        r->otel->parent_id   == NULL ||
+        r->otel->trace_flags == NULL)
+    {
+        goto error_state;
+    }
+
+    return NXT_OK;
+
+ error_state:
+    nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+    return NXT_ERROR;
+}
+
+
+nxt_int_t
+nxt_otel_parse_tracestate(void *ctx, nxt_http_field_t *field, uintptr_t data)
+{
+    nxt_http_request_t *r;
+    nxt_str_t     s;
+    nxt_http_field_t *f;
+
+    s.length = field->value_length;
+    s.start = field->value;
+    r = ctx;
+    r->otel->trace_state = s;
+
+    // maybe someday this should get sent down into the otel lib
+
+    f = nxt_list_add(r->resp.fields);
```

Comment:
It looks a bit strange to touch the response during request parse.
Does it also work with the following configuration?
```
{
    "action": {
          "return": 405
    }
}
```

Reply:
> It looks a bit strange to touch the response during request parse.

We. must do the header propagation. This is a core feature of opentelemetry. The Traceparent header may be generated when the span is created. This must then be added to the response.

Reply:
Note that in `nxt_http_error.c`, the `r->resp.fields` is reset. Does this meet the requirement?